### PR TITLE
JIT: Encode SIMD base type in VN for all HW intrinsics

### LINKED ISSUE #105721: JIT: VN does not encode base type for `CreateScalar`
```csharp
// Generated by Fuzzlyn v2.2 on 2024-07-30 18:05:36
// Run on Arm64 Windows
// Seed: 12131813856619591872-vectort,vector64,vector128,armsve
// Reduced from 251.5 KiB to 0.9 KiB in 00:03:25
// Debug: Outputs <-1, 0, 0, 0>
// Release: Outputs <65535, 0, 0, 0>
using System;
using System.Runtime.CompilerServices;
using System.Numerics;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.Arm;

public class C0
{
    public ulong F6;
}

public class C1
{
    public C0 F1;
    public C1(C0 f1)
    {
        F1 = f1;
    }
}

public class Program
{
    public static short s_21;
    public static C1 s_34 = new C1(new C0());
    public static void Main()
    {
        C0 vr7 = s_34.F1;
        vr7.F6 = (ulong)s_21--;
        var vr8 = (short)1;
        var vr9 = Vector128.CreateScalar(vr8).AsVector();
        var vr10 = Vector128.CreateScalar(s_21).AsVector();
        if (!!Sve.TestFirstTrue(vr9, vr10))
        {
            var vr11 = (int)s_21;
            Vector<int> vr12 = Vector128.CreateScalar(vr11).AsVector();
            System.Console.WriteLine(vr12);
        }
    }
}
```

cc @dotnet/jit-contrib @dotnet/arm64-contrib 

(This one may potentially not be SVE related -- not sure.)

### PR DESCRIPTION
**Pull request by jakobbotsch**
This changes the VN funcs for all HW intrinsic functions to always have an extra parameter for the SIMD base type. Before this change it was based on the instruction list, however that is not always the right thing (e.g. #105721 is a bug because of that). We also kept it as part of the HW intrinsic table, but that requires manual maintenance and is easy to get wrong. Always encoding the type is much more simple and the diffs do not look too bad.

In .NET 10 we can decide if we want to opt some intrinsics into not being differentiated based on the SIMD base type. The easiest thing might be to always map those to have the same base SIMD type (e.g. `CORINFO_TYPE_BYTE`) so that we don't end up with differences in arities for some VN functions that are hard to reason about.

Fix #105721

### CODE REVIEW

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @tannergooding 

I went with your suggestion to just encode the SIMD base type for all HW intrinsics, and removed all the references to the old stuff. The [diffs (from the run before I pushed the test)](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764060&view=ms.vss-build-web.run-extensions-tab) are miniscule so definitely seems like the right direction to go here.

**Comment by tannergooding** [comment]
> The [diffs (from the run before I pushed the test)](https://dev.azure.com/dnceng-public/public/_build/results?buildId=764060&view=ms.vss-build-web.run-extensions-tab) are miniscule so definitely seems like the right direction to go here.

👍, mostly looks to be around bitwise operations (and, or, xor, not, ternlog, etc). Users have the ability to manually workaround for these if its important in .NET 9 and mixing types to that degree is relatively rare, so I wouldn't expect most code to be pessimized by this. Should be easy to improve the select cases in .NET 10.

File: src/coreclr/jit/valuenum.cpp [hunk 1]
```diff
@@ -13049,22 +12990,12 @@ void Compiler::fgValueNumberHWIntrinsic(GenTreeHWIntrinsic* tree)
...
-                // There are zero arg HWINTRINSICS operations that encode the result type, i.e.  Vector128_AllBitSet
-                normalPair = vnStore->VNPairForFunc(tree->TypeGet(), func, resultTypeVNPair);
-                assert(vnStore->VNFuncArity(func) == 1);
-            }
-            else
-            {
-                normalPair = vnStore->VNPairForFunc(tree->TypeGet(), func);
-                assert(vnStore->VNFuncArity(func) == 0);
-            }
+            // There are zero arg HWINTRINSICS operations that encode the result type, i.e.  Vector128_AllBitSet
```
Comment (tannergooding) [inline src/coreclr/jit/valuenum.cpp:12996]: This comment is notably out of date, we produce `GT_CNS_VEC` for cases like `AllBitsSet` now. Off the top of my head, other zero arg intrinsics should also be getting converted to constant nodes and I'm not sure any cases actually exist that can hit this anymore.

Should be safe to leave in for .NET 9, but might be something we can cleanup early in .NET 10

**Review by tannergooding — APPROVED** [review]

**Comment by a74nh** [comment]
We're getting failures on all the SVE tests. We suspect it's this PR, but just confirming...

```
  Starting test: /home/mikabl01/dotnet/runtime/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false -p System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true ./artifacts/tests/coreclr/linux.arm64.Checked/JIT/HardwareInt  rinsics/HardwareIntrinsics_Arm_ro/HardwareIntrinsics_Arm_ro.dll GatherVectorFirstFaulting                                                                                                                                                                                                                                   
  ===================Running default===================                                                                                                                                                                                                                                                                       
  ------------------- {} -------------------                                                                                                                                                                                                                                                                                  
  Errors:                                                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                              
  Assert failure(PID 3341648 [0x0032fd50], Thread: 3341648 [0x32fd50]): Assertion failed 'vnStore->VNFuncArity(func) == 1' in 'JIT.HardwareIntrinsics.Arm._Sve.SveGatherVectorVectorBasesTest__Sve_GatherVectorFirstFaulting_Bases_double_ulong:RunBasicScenario_UnsafeRead():this' during 'Do value numbering' (IL size 110;   hash 0x04f13b09; FullOpts)                                                                                                                                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                              
      File: /home/mikabl01/dotnet/runtime/src/coreclr/jit/valuenum.cpp:12998                                                                                                                                                                                                                                                  
      Image: /home/mikabl01/dotnet/runtime/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/corerun   
```

**Comment by jakobbotsch** [comment]
> We're getting failures on all the SVE tests. We suspect it's this PR, but just confirming...
> 
> ```
>   Starting test: /home/mikabl01/dotnet/runtime/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false -p System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization=true ./artifacts/tests/coreclr/linux.arm64.Checked/JIT/HardwareInt  rinsics/HardwareIntrinsics_Arm_ro/HardwareIntrinsics_Arm_ro.dll GatherVectorFirstFaulting                                                                                                                                                                                                                                   
>   ===================Running default===================                                                                                                                                                                                                                                                                       
>   ------------------- {} -------------------                                                                                                                                                                                                                                                                                  
>   Errors:                                                                                                                                                                                                                                                                                                                     
>                                                                                                                                                                                                                                                                                                                               
>   Assert failure(PID 3341648 [0x0032fd50], Thread: 3341648 [0x32fd50]): Assertion failed 'vnStore->VNFuncArity(func) == 1' in 'JIT.HardwareIntrinsics.Arm._Sve.SveGatherVectorVectorBasesTest__Sve_GatherVectorFirstFaulting_Bases_double_ulong:RunBasicScenario_UnsafeRead():this' during 'Do value numbering' (IL size 110;   hash 0x04f13b09; FullOpts)                                                                                                                                                                                                                                                                                                  
>                                                                                                                                                                                                                                                                                                                               
>       File: /home/mikabl01/dotnet/runtime/src/coreclr/jit/valuenum.cpp:12998                                                                                                                                                                                                                                                  
>       Image: /home/mikabl01/dotnet/runtime/artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root/corerun   
> ```

Seems likely, sorry about that. How do I run those tests?

**Comment by jakobbotsch** [comment]
Although, the failing test was presumably added in #105595, so it might be an unlucky conflict with this PR (this PR went in shortly before that one).

**Comment by SwapnilGaikwad** [comment]
Hi @jakobbotsch , the same issue is also observed for other intrinsics. I checked the `Multiply`, `AddSaturate`, `SubtractSaturate`. 
You can run tests for the above intrinsics using the following steps.
1. Build the tests -  `./src/tests/build.sh checked -test:JIT/HardwareIntrinsics/HardwareIntrinsics_Arm_ro.csproj` (Windows equivalent of the command)
2. Run the tests - 
```
DOTNET_TieredCompilation=0 $CORE_ROOT/corerun ./artifacts/tests/coreclr/linux.arm64.Checked/JIT/HardwareIntrinsics/HardwareIntrinsics_Arm_ro/HardwareIntrinsics_Arm_ro.dll Sve_Multiply
```
Here, `CORE_ROOT=artifacts/tests/coreclr/linux.arm64.Checked/Tests/Core_Root`

**Pull request merged**